component/connector.go

```diff
@@ -0,0 +1,484 @@
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// TODO thorough explanation of connectors
+type Connector interface {
+	Component
+}
+
+// A TracesToTracesConnector sends traces from one pipeline to another.
+// Its purpose is to allow for differentiated processing of traces.
+// TracesToTracesConnector feeds a consumer.Traces with data.
+//
+// For example traces could be collected in one pipeline and routed to another traces pipeline
+// based on criteria such as attributes or other content of the trace. The second pipeline can
+// then process and export the trace to the appropriate backend.
+type TracesToTracesConnector interface {
+	Connector
+	ConsumeTracesToTraces(ctx context.Context, td ptrace.Traces) error
+}
+
+// A TracesToMetricsConnector acts as an exporter from a traces pipeline and a receiver to a metrics pipeline.
+// Its purpose is to derive metrics from a traces pipeline.
+// TracesToMetricsConnector feeds a consumer.Metrics with data.
+//
+// For example traces could be summarized by a metrics connector that emits statistics describing the traces observed.
+type TracesToMetricsConnector interface {
+	Connector
+	ConsumeTracesToMetrics(ctx context.Context, td ptrace.Traces) error
+}
+
+// A TracesToLogsConnector acts as an exporter from a traces pipeline and a receiver to a logs pipeline.
+// Its purpose is to derive logs from a traces pipeline.
+// TracesToLogsConnector feeds a consumer.Logs with data.
+//
+// For example traces could be analyzed by a logs connector that emits events when particular criteria are met.
+type TracesToLogsConnector interface {
+	Connector
+	ConsumeTracesToLogs(ctx context.Context, td ptrace.Traces) error
+}
+
+// A MetricsToTracesConnector acts as an exporter from a metrics pipeline and a receiver to a traces pipeline.
+// Its purpose is to derive traces from a metrics pipeline.
+// MetricsToTracesConnector feeds a consumer.Traces with data.
+//
+// For example latency between related data points could be modeled and emitted as traces.
+type MetricsToTracesConnector interface {
+	Connector
+	ConsumeMetricsToTraces(ctx context.Context, tm pmetric.Metrics) error
+}
+
+// A MetricsToMetricsConnector sends metrics from one pipeline to another.
+// Its purpose is to allow for differentiated processing of metrics.
+// MetricsToMetricsConnector feeds a consumer.Metrics with data.
+//
+// For example metrics could be collected in one pipeline and routed to another metrics pipeline
+// based on criteria such as attributes or other content of the metric. The second pipeline can
+// then process and export the metric to the appropriate backend.
+type MetricsToMetricsConnector interface {
+	Connector
+	ConsumeMetricsToMetrics(ctx context.Context, tm pmetric.Metrics) error
```

Comment:
Why do we need new func (`ConsumeMetricsToMetrics`) and just implementing `consumer.Metrics` is not enough? Signature is equivalent.

Reply:
This was my original implementation but I found it to be confusing in some ways. I don't think there's anything that technically prevents us from doing it as you suggested so I can switch it back if you want. 

To try to summarize the confusion I found though:

Conceptually, any connector that consumes traces is a `consumer.Traces`, regardless of what it does with them. We also need each combination to implement a different interface, and as you pointed out `consumer.Traces` can differentiate `traces->traces` from the other combinations. However, the others ultimately must be wrapped in `consumer.Traces` as well in order to allow other components to send data to them. I found this confusing because it means that `traces->traces` is always `consumer.Traces`, but `traces->metrics` is only sometimes `consumer.Traces`, depending on where in the code. I found it easier to just treat all combinations the same so we can say they each have a dedicated interface, and so they all become `consumer.Traces` at the same time when building the graph. 

I'm not certain, but from a component author's point of view, the same confusion may be relevant.

Reply:
Can you give me a bit of code snippet to make sure we understand. I am a bit confused about the explanation.

Reply:
I'll try to explain better. It may be necessary to first step back and highlight a couple design questions that I worked through in this implementation.
- What constraints should be placed on the input->output types supported by a connector type?
- Based on the input/output constraints, how do users reason about configuration?
- Based on the input/output constraints, how should connectors be managed within the collector - instanced per-pipeline like processors, shared per-signal like receivers/exporters, or something else?

---

> What constraints should be placed on the input->output types supported by a connector type?

I considered the following options:
1. A connector type can support _exactly one input/output combination_. This is too granular and prevents intuitive use cases like `nop` and `count`, instead requiring `noptraces`, `nopmetrics`, etc.
2. A connector type can support _exactly one output type per input type_. This works for `nop` (`traces->traces, metrics->metrics, logs->logs`) and `count` (`traces->metrics, metrics->metrics, logs->logs`).
3. A connector type can support _any or all input/output combinations_. This is the least restrictive but introduces some implementation considerations. Notably though, this. is necessary to directly support behavior like `spanmetricsprocessor`, which _both_ derives metrics from spans _and_ allows the original spans to pass through unmodified. (i.e.`traces->traces, traces->metrics`).

I think option 3 is best because it allows for the most use cases / is the least restrictive.

---

> Based on the input/output constraints, how do users reason about configuration?

In short, a connector will do all the things it can do that are applicable to the pipelines it connects.
- A connector type may consume one or more signal types and emit one or more signal types.
- Each connector type declares support for specific combinations of signal types. 
- Each signal combination supported by a connector is an independent function. That is, the connector does not require that you use all if its capabilities. For example, if the `count` connector is configured to connect from a logs pipeline to a metrics pipeline, it will count logs. Likewise, if you do not want the `spanmetrics` connector to forward traces, you can configure only the `traces->metrics` connection and ignore the `traces->traces` capability.
- A connector will consume/emit for each pair of pipelines that it is connected.
- If a connector configured in such a way that any combination of its input/output pipelines are unsupported, this will result in a configuration error.


---

> Based on the input/output constraints, how should connectors be managed within the collector - instanced per-pipeline like processors, shared per-signal like receivers/exporters, or something else?

A connector is instanced based on each input/output type. This is similar to how receivers and exporters are managed internally. Specially, receivers and exporters are instanced per signal type. (Technically, the component developer can use `sharedcomponent` to unify these instances, but the collector manages them via a separate handle for each signal type.)

[This example](https://github.com/djaglowski/opentelemetry-collector/blob/connectors-prototype-v2/service/internal/pipelines/testdata/pipelines_conn_matrix.yaml) shows a connector that is consuming and emitting all three signal types. Internally, the collector will instantiate 9 instances of this connector, with each instance constructed by the corresponding `Create<DataType>To<DataType>Connector` factory method. Likewise, each instance of the connector will consume signals by performing the corresponding `Consume<DataType>To<DataType>` function.

In such a case, each of the 9 instances must ultimately satisfy one of the `consumer.<DataType>` interfaces. This is done by wrapping each instances via the appropriate `consumer.New<DataType>` function. [Here is the code that handles this](https://github.com/djaglowski/opentelemetry-collector/blob/244af4ce8c9ddc7fc435b7f25417298e0a4552b4/service/internal/pipelines/nodes.go#L299), but as a simplified version:

```go
switch receiverPipelineType {
case component.DataTypeTraces:
	...
	switch exporterPipelineType {
	case component.DataTypeTraces:
		connectorInstance := n.factory.CreateTracesToTracesConnector(ctx, set, cfg, fanoutConsumer)
		tracesConsumer := consumer.NewTraces(connectorInstance.ConsumeTracesToTraces)
		...
	case component.DataTypeMetrics:
		connectorInstance := n.factory.CreateMetricsToTracesConnector(ctx, set, cfg, fanoutConsumer)	
		metricsConsumer := consumer.NewMetrics(connectorInstance.ConsumeMetricsToTraces)
		...
	case component.DataTypeLogs:
		connectorInstance := n.factory.CreateLogsToTracesConnector(ctx, set, cfg, fanoutConsumer)
		logsConsumer := consumer.NewLogs(connectorInstance.ConsumeLogsToTraces)
		...
	}
```

We _could_ rename `ConsumeTracesToTraces` to `ConsumeTraces`, but then it is a special case vs `ConsumeMetricsToTraces` and `ConsumeLogsToTraces`. This is harder to keep track of in my opinion. It becomes especially awkward when trying to unwrap the connector instance, such as is necessary in tests when validating that each component in the graph is as expected. By treating each signal combination as equal, it allows for uniform code [such as this](https://github.com/djaglowski/opentelemetry-collector/blob/244af4ce8c9ddc7fc435b7f25417298e0a4552b4/service/internal/pipelines/graph_test.go#L108-L118).

Reply:
@djaglowski this requires every pairwise combination of signals to have an interface. Is this approach going to scale if we add one or two other new signals? It is 3x3 now, if we add 2 more signals (e.g. profiles and entities), we will have 5x5=25 interfaces, 25 methods and 25 switch cases, etc.
Is there a way to use generics here somehow to reduce the code volume? (Probably no, I don't remember Go's genics having much compile-time power to generate combinations of type parameters).

Reply:
25 combinations seems... manageable, if far from ideal. The scalability issues is _somewhat_ mitigated in that we likely will not exceed 5 signal types any time soon. I agree though - the current implementation will become quite cumbersome when adding additional signals.

I'll spend some time looking into this. There might be a way to decouple the input/output types so that a connector has to implement one consumer type and one "emitter" type, and then rely on composition rather than pairwise combinations.

Reply:
I'll play around with this a bit, but I believe it might be possible to use generics here to alleviate at least part of the combinatorial problem.